File: Data Structures & Algorithms/products-of-array-discluding-self/submission-7.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> result = {1};
        for (int i = 1; i < nums.size(); i++) {
            int n = nums[i - 1];
            result.emplace_back(n * result[i - 1]);
        }
        int accumulate = 1;
        for (int i = nums.size() - 1; i > -1; i--) {
            result[i] = result[i] * accumulate;
            accumulate = accumulate * nums[i];
        }
        return result;
    }
};

File: Data Structures & Algorithms/maximum-product-subarray/submission-7.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        Min, Max = 1, 1
        result = float("-inf")
        for n in nums:
            temp = Max
            Max = max(temp * n, Min * n, n)
            Min = min(temp * n, Min * n, n)
            result = max(result, Max)
        return result